pallets/foreign-investments/src/lib.rs

```diff
@@ -43,43 +43,137 @@
-/// <https://docs.substrate.io/reference/frame-pallets/>
+use cfg_traits::{investments::TrancheCurrency, PoolInspect};
+use cfg_types::investments::{CollectedAmount, Swap};
+use frame_support::{dispatch::DispatchResult, ensure};
 pub use pallet::*;
-
-pub mod errors;
-pub mod hooks;
-pub mod impls;
-pub mod types;
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_runtime::{traits::EnsureSub, ArithmeticError, DispatchError};
 
 #[cfg(test)]
 mod mock;
 
 #[cfg(test)]
 mod tests;
 
+/// Hold the base information of a foreign investment/redemption
+#[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct BaseInfo<T: Config> {
+	foreign_currency: T::CurrencyId,
+	pool_currency: T::CurrencyId,
+	collected: CollectedAmount<T::Balance>,
+}
+
+impl<T: Config> BaseInfo<T> {
+	fn new(
+		investment_id: T::InvestmentId,
+		foreign_currency: T::CurrencyId,
+	) -> Result<Self, DispatchError> {
+		Ok(Self {
+			foreign_currency,
+			pool_currency: T::PoolInspect::currency_for(investment_id.of_pool())
+				.ok_or(Error::<T>::PoolNotFound)?,
+			collected: CollectedAmount::default(),
+		})
+	}
+
+	fn ensure_same_foreign(&self, foreign_currency: T::CurrencyId) -> DispatchResult {
+		ensure!(
+			self.foreign_currency == foreign_currency,
+			Error::<T>::MismatchedForeignCurrency
+		);
+
+		Ok(())
+	}
+}
+
+/// Hold the information of a foreign investment
+#[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct InvestmentInfo<T: Config> {
+	base: BaseInfo<T>,
+	increase_swap_id: Option<T::SwapId>,
+	decrease_swap_id: Option<T::SwapId>,
+	total_pool_amount: T::Balance,
+}
+
+impl<T: Config> InvestmentInfo<T> {
+	fn new(
+		investment_id: T::InvestmentId,
+		foreign_currency: T::CurrencyId,
+	) -> Result<Self, DispatchError> {
+		Ok(Self {
+			base: BaseInfo::new(investment_id, foreign_currency)?,
+			total_pool_amount: T::Balance::default(),
+			increase_swap_id: None,
+			decrease_swap_id: None,
+		})
+	}
+
+	fn remaining_pool_amount(&self) -> Result<T::Balance, ArithmeticError> {
+		self.total_pool_amount
+			.ensure_sub(self.base.collected.amount_payment)
+	}
+}
+
+/// Hold the information of an foreign redemption
+#[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct RedemptionInfo<T: Config> {
+	base: BaseInfo<T>,
+	swap_id: Option<T::SwapId>,
+	total_tranche_tokens: T::Balance,
+	swapped_amount: T::Balance,
+	tranche_tokens_sent: T::Balance,
+}
+
+impl<T: Config> RedemptionInfo<T> {
+	fn new(
+		investment_id: T::InvestmentId,
+		foreign_currency: T::CurrencyId,
+	) -> Result<Self, DispatchError> {
+		Ok(Self {
+			base: BaseInfo::new(investment_id, foreign_currency)?,
+			swap_id: None,
+			total_tranche_tokens: T::Balance::default(),
```

Comment:
What does `total` here refer to? The amount of tranche tokens which are pending for fulfillment?

Reply:
It refers to the number of tranche tokens the user has decided to redeem via increase/decrease. The expected amount. Nothing to do with collected or swapped.

Reply:
Maybe `total_` to `expected_` clarifies this?

Reply:
I propose pending because they are `pending` for redemption. That also aligns with the language we settled with for the `InvestmentPortfolio`.

Reply:
The "problem" I see with `pending` is that it can be confused with `pending` to collect or `pending` to swap. Actually, I think it's not `pending` as it works right now because after the user redeems part of it, this variable does not change.

Let me consider if I can rename/refactor this to better understand it 🤔

Reply:
Refactored here [38347ff](https://github.com/centrifuge/centrifuge-chain/pull/1698/commits/38347ff110bd287b652bddea40e0eed147f4c121) following your suggestion